draft-omara-sframe.md

```diff
@@ -372,91 +266,137 @@ Frame counter (CTR): (Variable length)
+  opposed to a separate field as in SRTP {{?RFC3711}}).
 
-2- Encryption key to encrypt the media frame
-
-~~~~~
-Key = HKDF(K, 'SFrameEncryptionKey', 16)
-~~~~~
+* `AEAD.Nk` - The size of a key for the encryption algorithm, in bytes
 
-3- Authentication key to authenticate the encrypted frame and the media metadata
+* `AEAD.Nn` - The size of a nonce for the encryption algorithm, in bytes
 
-~~~~~
-Key = HKDF(K, 'SFrameAuthenticationKey', 32)
-~~~~~
+### Key Derivation
 
+Each client creates a secret key `base\_key` and shares it with the other
+participants via an E2E secure channel.  From `base\_key`, we derive the
+following values using HKDF {{!RFC5869}}:
 
-The IV is 128 bits long and calculated from the CTR field of the Frame header. First the var int CTR is decoded into 16 bytes long field, where CTR fills the 
-the 8 most significant bytes and low 8 bytes are always filled with zeros:
+1. A Salt key used to calculate the IV
+2. An encryption key to encrypt the media frame
+3. An authentication key to authenticate the encrypted frame and the media metadata
 
 ~~~~~
-CTR16 = DecodeVarInt(CTR, 16)
-IV = CTR16 XOR Salt key
+sframe_secret = HKDF-Extract(K, 'SFrame10')
+sframe_key = HKDF-Expand(sframe_secret, 'key', AEAD.Nk)
+sframe_salt = HKDF-Expand(sframe_secret, 'nonce', AEAD.Nn)
 ~~~~~
 
+Each SFrame KID value identifies a `base\_key`, and thus both the
+`sframe\_secret` and `sframe\_salt` values to be used for encryption and
+decryption.
 
 ### Encryption
+
 After encoding the frame and before packetizing it, the necessary media metadata will be moved out of the encoded frame buffer, to be used later in the RTP generic frame header extension. The encoded frame, the metadata buffer and the frame counter are passed to SFrame encryptor.
-The encryptor constructs SFrame header using frame counter and key id and derive the encryption IV. The frame is encrypted using the encryption key and the header, encrypted frame, the media metadata and the header are authenticated using the authentication key. The authentication tag is then truncated (If supported by the cipher suite) and prepended at the end of the ciphertext.
 
-The encrypted payload is then passed to a generic RTP packetized to construct the RTP packets and encrypts it using SRTP keys for the HBH encryption to the media server.
+SFrame encryption uses the AEAD encryption algorithm for the ciphersuite in use.
+The key for the encryption is the `sframe\_key` and the nonce is formed by XORing
+the `sframe\_salt` with the current counter, encoded as a big-endian integer of
+length `AEAD.Nn`.
+
+The encryptor forms an SFrame header using the S, CTR, and KID values provided.
+The encoded header is provided as AAD to the AEAD encryption operation, with any
+frame metadata appended.
+
+~~~~~
+def encrypt(S, CTR, KID, frame_metadata, frame):
+  sframe_key, sframe_salt = key_store[KID]
+
+  frame_ctr = encode_big_endian(CTR, AEAD.Nn)
```

Comment:
Neel to spell out explicitly that the ctr is encoded into 16 bytes buffer, where the ctr occupies the left 8 bytes and padded with 8 zero bytes

Reply:
This is a bit more flexible than that, in order to allow for cipher agility.  CTR and GCM take 12-byte IVs, whereas say CBC takes 16 bytes and XChaCha20 takes 24 (not that we would necessarily use those, but they're not totally implausible).  So what this is mean to say is, "Encode `CTR` into a big-endian integer as long as a nonce", which means implicitly that the number is on the right, padded with zeros to the left.

```
encode_big_endian(0x123456789ab, 12) = 0x0000000000000123456789ab
encode_big_endian(0x123456789ab, 16) = 0x000000000000000000000123456789ab
```

Unless you disagree with forming the nonce that way, the right answer is probably just to properly define `encode_big_endian()`.